### PR TITLE
[stdlib] Nest Iterator and Index types for various stdlib types

### PR DESCRIPTION
**Pull request by airspeedswift**
Nests various custom iterators and indexes for stdlib types inside their used types. This has been possible for a while now that we have nested generics, and cleans up the namespace.

Should be a non-functional-change for now with type aliases for the old names which aren't yet deprecated (deprecation/removal of type aliases is not ABI impacting so can be done at any time).

Ignoring `Range` and `String` types for now since they're in flux.

### CODE REVIEW

**Comment by airspeedswift** [comment]
@swift-ci please smoke benchmark

**Comment by airspeedswift** [comment]
@swift-ci please test source compatibility

**Comment by airspeedswift** [comment]
@swift-ci please smoke test compiler performance

**Comment by airspeedswift** [comment]
@swift-ci please test

**Comment by airspeedswift** [comment]
@swift-ci please test

**Comment by airspeedswift** [comment]
@swift-ci please test

**Comment by airspeedswift** [comment]
@swift-ci please smoke test compiler performance

**Comment by airspeedswift** [comment]
@swift-ci please smoke test compiler performance

**Comment by airspeedswift** [comment]
@shahmishal @graydon any ideas why the compiler perf tests are failing? Error in the logs says:

```
suite/project_future.py", line 836, in output_fd
    ([scheme_target] if scheme_target else []) +
TypeError: string indices must be integers
```

**Comment by airspeedswift** [comment]
@swift-ci please test source compatibility

**Comment by airspeedswift** [comment]
@swift-ci please test source compatibility

**Comment by slavapestov** [comment]
Neat!

File: stdlib/public/core/EmptyCollection.swift [hunk 1]
```diff
@@ -17,32 +17,42 @@
...
   /// Creates an instance.
   @_inlineable // FIXME(sil-serialize-all)
   public init() {}
 }
 
-extension EmptyIterator: IteratorProtocol, Sequence {
+extension EmptyCollection {
+  /// An iterator that never produces an element.
+  @_fixed_layout // FIXME(sil-serialize-all)
+  public struct Iterator {
```
Comment (natecook1000) [inline stdlib/public/core/EmptyCollection.swift]: Could `EmptyCollection` just be its own `Iterator`?

Comment (airspeedswift) [reply stdlib/public/core/EmptyCollection.swift]: I guess. I think I prefer it consistent tho.

**Comment by airspeedswift** [comment]
@swift-ci please smoke test compiler performance

**Comment by DougGregor** [comment]
@airspeedswift , this caused the regression in https://bugs.swift.org/browse/SR-6654

**Pull request merged**